### PR TITLE
chore(export_scene_macro): fixes and improvements

### PR DESCRIPTION
**Pull request by dasmy**
### Context
Although ParaView supports exporting a scene to `*.vtkjs`, the `export_scene_macro` is still useful for rapid customization, prototyping, and educational purposes.

One of the examples is https://discourse.paraview.org/t/trouble-exporting-scene-to-vtkjs/9829/10, where the native export was incomplete, but fixing it in C++ was infeasible for me. This is why I looked at the macro, update
[SceneExplorer_node-18.x.html.zip](https://github.com/Kitware/vtk-js/files/9050061/SceneExplorer_node-18.x.html.zip)
d it to support Python 3, fixed some issues and added the features that were missing for my use case.
This PR also fixes https://github.com/Kitware/vtk-js/issues/2499 which I found while looking at the code.

### Results
* The macro runs again (didn't any more, tested in ParaView 5.10.1 on macOS)
* The macro does not only export arrays relevant for coloring but all existing arrays (and in contrast to the ParaView native export also `vtkStringArray`).

### Changes
The following changes have been made (see also git history):
* Fix errors (Python 3 compatibility, name of imported modules)
* Do not ignore vtkMultiBlockDataset when exporting
* Export all numeric arrays instead only the one needed for coloring the scene items
* Do not skip vtkStringArray when exporting to *.vtkjs
* Fixed export when names of scene items contain spaces or other special characters
* Support for coloring of scene items by non-numeric arrays (tested with string arrays)
* Python-style formatting

### Testing
- Run the macro from ParaView and inspect the produced `*.vtkjs` to see that all array data is now available.
- The exported scene can also be analyzed using a modified SceneExplorer, see attachment (from https://github.com/dasmy/SceneExplorer), where labels based on numeric and string-based arrays can be added to point data visualizations.
image

- Tested environment:
  - **vtk.js**: not relevant
  - **OS**: macOS
  - **Browser**: not relevant
  - **Paraview**: 5.10.1

### CODE REVIEW

**Comment by dasmy** [comment]
Btw.: I would love to see some proper output on where the exported file has been saved, but actually all `print()` statements are ignored if the Python console is closed. Is there any way of showing a message box (or even better: opening a file selection dialog so that the user can select where to store the output)?

**Comment by finetjul** [comment]
Can you please revert indent spacing to 2 spaces instead of 4 ?

File: Utilities/ParaView/export-scene-macro.py [hunk 1]
```diff
@@ -430,12 +440,13 @@ def mkdir_p(path):
...
       continue
+
     if hasattr(renProp, 'GetMapper'):
       mapper = renProp.GetMapper()
-      dataObject = mapper.GetInputDataObject(0, 0);
+      dataObject = mapper.GetInputDataObject(0, 0)
       dataset = None
 
-      if dataObject.IsA('vtkCompositeDataSet'):
+      if dataObject.IsA('vtkCompositeDataSet') or dataObject.IsA('vtkMultiBlockDataSet'):
```
Comment (jourdain) [inline Utilities/ParaView/export-scene-macro.py:449]: `vtkMultiBlockDataSet` is actually a `vtkCompositeDataSet` so the previous condition should work just fine.

**Comment by jourdain** [comment]
the code looks good. Ideally we may want to re-work the commit message to follow semantic-release/commitizen while not triggering any new release as this code is only available here as reference.

**Comment by jourdain** [comment]
```
0s
Run npx commitlint --from=[1](https://github.com/Kitware/vtk-js/runs/7212189718?check_suite_focus=true#step:6:1)[4](https://github.com/Kitware/vtk-js/runs/7212189718?check_suite_focus=true#step:6:5)aff403bdeba9edd4bc5b4e4ef911abf3227f9e
⧗   input: autopep8 formatting:
autopep8 --in-place --indent-size=2 --ignore=E121 export-scene-macro.py
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
⧗   input: also supoprt coloring by a non-numeric array
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
⧗   input: fix export of items with spaces (and other special chars) in their name
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
⧗   input: support exporting string arrays
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
⧗   input: export all standard arrays instead of only the color
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
⧗   input: fix errors, also export vtkMultiBlockDataSet
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
```

**Comment by dasmy** [comment]
> Also, I suggest replacing the mutable default values with `None`, i.e. replace all the `root={}` with `root=None` in the function definitions and add the lines:
> 
> ```python
> if root is None:
>   root = {}
> ```
> 
> (As explanation: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)

Definitely agree. Fixed it. Originally I did not want to touch to many things but as the change is a bit larger now anyway, this additional change does not make much of a difference :-D

Actually, the interface design of these routines seems bad anyway. From a first look, I would assume it might be better to drop the argument, always create a local root inside the routines and return it. Then, a caller could integrate the subtree it just received into its own local root. But didn't look very much in detail and would rather not rewrite the whole thing even more...

**Comment by dasmy** [comment]
> Thanks for the commit rewrites but `feat()` and `fix()` will trigger new releases. To prevent any release, you have to stick with `style()` and `chore()`. Sorry for not mentioning before.

Ooops - didn't know that (and didn't read it anywhere). - Changed the messages again.

**Review by jourdain — APPROVED** [review]
LGTM

**Pull request merged**